Use mysql on insert duplicate update query to reduce number of aggregation query per course.

taranjeet opened the pull request:
**Description:** 

This PR adds support to reduce the number of queries made to insert aggregation data per course. It uses Mysql Insert on Duplicate Update query, so that a single query can be sent which contains manyu rows of aggregation data

**JIRA:** [OC-4844](https://tasks.opencraft.com/browse/OC-4844)

**Dependencies:** NA

**Merge deadline:** List merge deadline (if any)

**Installation instructions:** List any non-trivial installation 
instructions.

**Testing instructions:**

Tests should pass

**Reviewers:**
- [ ] @symbolist 
- [ ] @jcdyer 

**Merge checklist:**
- [ ] All reviewers approved
- [ ] CI build is green
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Commits are squashed
- [ ] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPI after tag-triggered build is 
      finished.
- [ ] Delete working branch (if not needed anymore)

**Author concerns:** 
* Functions like `AggregatorManager.validate` and the logic to calculate percentage completion are duplicated in `AggregationUpdater._is_aggregator_data_valid` and `utils.get_percent`. I was not sure where to keep them or convert the existing method to static method or so.

taranjeet commented:
@jcdyer @symbolist : this is ready for review :)

symbolist commented on completion_aggregator/tasks/aggregation_tasks.py:
```diff
@@ -201,18 +220,47 @@ def update_for_aggregator(self, block, affected_aggregators, force):
+                'block_key': block,
+                'aggregation_name': block.block_type
+            })
+            # Param order: (user_id, course_key, block_key, block_type, earned, possible, percent,
```
I know this pattern is used in the file but not sure I like it. It is bypassing model level validation, the internal model details are being exposed here and it cannot be tested in isolation. How about this:

1. We create a Aggregator instance (without saving it) and add it to the list.
2. The Aggregator model can have a method `values()` which will return a dictionary of values of its fields. Similar to https://docs.djangoproject.com/en/1.11/ref/models/querysets/#values. 
3. It can be called in `save_updated_aggregators()` to collect the data from the model instances.
4. In the query, can we not use keyed arguments rather than positional? https://docs.djangoproject.com/en/1.11/topics/db/sql/#passing-parameters-into-raw

taranjeet replied:
I have given this a try. Maybe variable names need to be updated a bit.

symbolist commented on completion_aggregator/tasks/aggregation_tasks.py:
```diff
@@ -201,18 +220,47 @@ def update_for_aggregator(self, block, affected_aggregators, force):
+            ])
         return CompletionStats(earned=total_earned, possible=total_possible, last_modified=last_modified)
 
+    def update_aggregator_at_once(self):
```
Actually, how about moving this to the `AggregatorManager` and calling it something like `bulk_create_or_update(objects)` similar to https://docs.djangoproject.com/en/1.11/ref/models/querysets/#bulk-create?

taranjeet replied:
This is fixed.

symbolist commented on completion_aggregator/tasks/aggregation_tasks.py:
```diff
@@ -245,6 +293,43 @@ def _aggregator_needs_update(self, block, modified, force):
             return getattr(agg, 'last_modified', OLD_DATETIME) < modified
         return False
 
+    def _is_aggregator_data_valid(self, user, course_key, block_key):
```
Okay, just saw your comment in the description. It does not need to be converted to a class method since a manager instance is available at `Aggregator.objects`.

symbolist commented on completion_aggregator/utils.py:
```diff
@@ -28,3 +32,26 @@ def get_active_users(course_key):
+    return date
+
+
+def get_percent(earned, possible):
```
Re: your comment in description

It is fine place for it.

taranjeet replied:
This is fixed.

taranjeet commented on completion_aggregator/tasks/aggregation_tasks.py:
```diff
@@ -201,18 +220,47 @@ def update_for_aggregator(self, block, affected_aggregators, force):
+            })
+            # Param order: (user_id, course_key, block_key, block_type, earned, possible, percent,
+            #               last_modified, created, updated)
+            self.aggregator_list.append([
```
This is resolved now.

taranjeet commented on completion_aggregator/tasks/aggregation_tasks.py:
```diff
@@ -201,18 +220,47 @@ def update_for_aggregator(self, block, affected_aggregators, force):
-                possible=total_possible,
-                last_modified=last_modified,
-            )
-            self.aggregators[obj.block_key] = obj
```
I have fixed this now.

taranjeet commented on completion_aggregator/tasks/aggregation_tasks.py:
```diff
@@ -201,18 +220,47 @@ def update_for_aggregator(self, block, affected_aggregators, force):
+                cur.executemany(INSERT_ON_DUPLICATE_UPDATE_QUERY, self.aggregator_list)
+            for qparams in self.processed_list:
+                obj = Aggregator.objects.get(**qparams)
+                self.aggregators[obj.block_key] = obj
```
You are right, This will be N get queries. 
So I have stored a map of `block_key : { 'last_modified': 'last_modified_value'}`, because all we are doing is accessing the value of last modified from this.

taranjeet commented:
@symbolist @jcdyer : This is ready for another pass. I have removed the code where I was querying individually for aggregator object. I am just storing the value of block key is to last modified. 

One thing to note here is that right now `self.aggregators` is a map which contains block key is to aggregator object. This can be updated to include an object of the form
```
{'last_modified': 'last_modified_value`}
``` 

once we are fine with the suggested approach.

jcdyer commented on completion_aggregator/utils.py:
```diff
@@ -28,3 +32,26 @@ def get_active_users(course_key):
+    # Reason: https://groups.google.com/forum/#!topic/django-developers/JA1uaAxnuGE
+    if DJANGO_MAJOR_VERSION >= 1 and DJANGO_MINOR_VERSION >= 10:
+        date = date.replace(tzinfo=None)
+    return date
```
My reading of the linked post is that raw `SELECT` queries are returning timezone unaware values, but I don't understand why we need to force versions to be unaware before submitting them.  I suspect this will cause problems when retrieving aggregators normally, and trying to compare them with normal aware datetimes.  Did you run into actual issues when implementing this?

taranjeet replied:
@jcdyer 

Yeah, I ran into this error. Here are version specific logs

**Django version 2.0** (requires me to pass a naive datetime)
```
django.utils.deprecation.RemovedInDjango20Warning: The MySQL database adapter received an aware datetime (%s), probably from cursor.execute(). Update your code to pass a naive datetime in the database connection's time zone (UTC by default).
```

**Django version 1.8** (requires me to pass an aware datetime)
```
RuntimeWarning: MySQL received a naive datetime (2018-08-08 09:46:22.687201) while time zone support is active.
```
Also here is the log from one of the [travis job](https://travis-ci.org/open-craft/openedx-completion-aggregator/jobs/413493947) when this code was not present.
 

My reason for linking to that blog post, is I could not find a proper link which give details about this. Hence I posted that link because it was giving context that timezone unaware values are being returned.

symbolist replied:
Hmm. Given:

1. https://docs.djangoproject.com/en/2.0/topics/i18n/timezones/#database
2. https://github.com/django/django/commit/e707e4c709c2e3f2dad69643eb838f87491891f8#diff-af003fcfed7cfbdeb396f8647ed0f92fR258
3. https://docs.djangoproject.com/en/2.1/ref/settings/#time-zone

I think we need to do `data.astimezone(timezone.utc).replace(tzinfo=None)` for Django 2+.

@jcdyer What do you think?

jcdyer replied:
Thanks for explaining @taranjeet.  That's helpful.  

@symbolist: The data *should* already be coming in with tzinfo=utc, but this would be safer, and should be a trivial amount of extra processing time. :+1:

symbolist commented:
@taranjeet Okay, the code is close to done now. You also need to write some tests to check all the new code. In the tests you can use [assertNumQueries](https://docs.djangoproject.com/en/1.11/topics/testing/tools/#django.test.TransactionTestCase.assertNumQueries) to verify the expected number of queries.

jcdyer commented on completion_aggregator/tasks/aggregation_tasks.py:
```diff
@@ -201,18 +207,35 @@ def update_for_aggregator(self, block, affected_aggregators, force):
                 earned=total_earned,
                 possible=total_possible,
-                last_modified=last_modified,
+                last_modified=last_modified
```
Yep.  The edx python style guide specifies using final commas (See the [third bullet point here](https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/python-guidelines.html#syntax-and-organization).  Fewer frustrations when rearranging or adding arguments.

taranjeet replied:
Also, it prevents the line being shown in `git diff` next time.

taranjeet commented:
@symbolist : I have addressed all the review comments and added test using `assertNumQueries` as well. Please have a review :)

symbolist commented on completion_aggregator/tasks/aggregation_tasks.py:
```diff
@@ -210,7 +216,8 @@ def update_for_aggregator(self, block, affected_aggregators, force):
             )
-            self.aggregators[obj.block_key] = obj
+            self.updated_aggregators.append(aggregator)
+            self.aggregators[block] = aggregator
```
I just realized that for blocks with an existing Aggregator object in `self.aggregators`, we should not be creating new objects but instead updating values of the existing objects (without saving them). Otherwise the `created` field will be reset.

symbolist replied:
Since `pre_save` is not going to get called, `modified` also needs to be updated.

taranjeet replied:
@symbolist : Please correct me if I am wrong.

I think this is why we are using "INSERT ON DUPLICATE UPDATE" query in mysql. If there will be any object(whose value is already present in the database) and it is being inserted, it will be updated with new values for `earned`, `total`, `percent` and `last_modified`. Another point to support this, when we use `submit_completion`, which internally uses `create_or_update`. It returns a flag which tells if the new object was created or not. In the present codebase, there is no support for such case.

Also since the `Aggregator` objects are not saved, `created` and `modified` will not be populated before being saved to db.

symbolist replied:
@taranjeet The thing is that `created` and `modified` fields have a default value set to the current time: https://github.com/jazzband/django-model-utils/blob/master/model_utils/fields.py#L12-L36.

Now, if I create a new Aggregator object for an existing `(user, course_key, block_key, aggregation_name)` object, its `created` and `modified` values are not going to be what they are in the database. So the Aggregator objects that are going to be passed to `bulk_create_or_update` and are going to be set in `self.aggregators` are going to be invalid. The difference with `create_or_update` is that the later is passed data with differentiation between the key fields and others and as you noted returns an object that can put into `self.aggregators`. Yes, this does work here because of how `bulk_create_or_update` is internally implemented but callers should not assume anything about that and because they are not currently looking at the `created` or `modified` fields.

As far as I see the only thing needed here is adding a check that if there is an existing object, four of its fields are updated. Otherwise a new object is created as is happening here already.

taranjeet replied:
@symbolist Thanks for the detailed info. :)

I have updated the codebase.

symbolist commented on completion_aggregator/models.py:
```diff
@@ -205,6 +225,21 @@ def __str__(self):
+        exact_keys = ['user', 'course_key', 'block_key', 'aggregation_name', 'earned', 'possible', 'last_modified']
+        values = {}
+        values = {key: getattr(self, key) for key in exact_keys}
+        values['user'] = values['user'].id
```
Why not `self.user.id` instead of putting it in `values` first?

symbolist commented on completion_aggregator/models.py:
```diff
@@ -205,6 +225,21 @@ def __str__(self):
+        values['last_modified'] = make_datetime_timezone_unaware(values['last_modified'])
+        values['percent'] = get_percent(values['earned'], values['possible'])
+        values['created'] = make_datetime_timezone_unaware(timezone.now())
+        values['modified'] = make_datetime_timezone_unaware(timezone.now())
```
We can't reset the value of created when updating existing `Aggregator` objects. In this method, we should only be reading the values for `self`. You can do instead:

```
values.update({key: make_datetime_timezone_unaware(getattr(self, key)) for key in [
   'last_modified', 'created', 'modified',
]})
```

taranjeet commented on completion_aggregator/models.py:
```diff
@@ -205,6 +225,21 @@ def __str__(self):
+        """
+        # these keys values will be assigned as it is
+        exact_keys = ['user', 'course_key', 'block_key', 'aggregation_name', 'earned', 'possible', 'last_modified']
+        values = {}
```
@symbolist Sorry, but I didn't understand this comment.

symbolist replied:
I meant this is not needed since it is being set in the next line.

taranjeet commented:
@symbolist : One more thing, I am not sure that we should write tests for `get_percent` and `get_values`, as they are invoked many times from code which is already being tested. So if anything breaks from these functions, all the tests will start failing.

Also, I was trying to see test cases for other model functions (like block_is_registered_aggregator) or util functions (like get_active_users), but I couldn't find any. Is there any specific reason why test cases are not present for these functions?

symbolist commented:
@taranjeet I do not have historical knowledge of this repo but these should be straightforward to add.

symbolist commented on completion_aggregator/tasks/aggregation_tasks.py:
```diff
@@ -201,16 +206,27 @@ def update_for_aggregator(self, block, affected_aggregators, force):
+
+            Aggregator.objects.validate(self.user, self.course_key, block)
+            aggregator_identifier_key = '$'.join(map(str, [self.user.id, self.course_key, block, block.block_type]))
+            if aggregator_identifier_key not in self.aggregator_map:
```
I don't see why this needs to be so complicated. You can just do `if block not in self.aggregators:`

taranjeet replied:
I thought I should use what makes an `Aggregator` object unique. Updating this to use `block` only.

symbolist commented on completion_aggregator/tasks/aggregation_tasks.py:
```diff
@@ -201,16 +206,27 @@ def update_for_aggregator(self, block, affected_aggregators, force):
+                aggregator.last_modified = last_modified
+            self.aggregator_map[aggregator_identifier_key] = aggregator
+            self.updated_aggregators.append(aggregator)
+            self.aggregators[block] = aggregator
```
This can move to the `if` since in the `else` it is already set.

taranjeet replied:
Got your point. Thanks.

taranjeet commented:
@symbolist : I have updated the pr as per latest set of review comments. It includes test cases for `get_percent`, `get_values` and a conditional clause to check if aggregator object is present, so its updated rather than being created.

The conditional clause one was a nice catch, which I was thinking that it will be handled totally by "INSERT ON DUP UPDATE" but your comment made more sense about the abstraction part. Thanks :)

taranjeet commented on completion_aggregator/tasks/aggregation_tasks.py:
```diff
@@ -201,16 +205,25 @@ def update_for_aggregator(self, block, affected_aggregators, force):
+                aggregator = self.aggregators[block]
+                aggregator.earned = total_earned
+                aggregator.possible = total_possible
+                aggregator.last_modified = last_modified
```
This is addressed. Thanks for pointing it out @symbolist :)

taranjeet commented:
@symbolist : I have addressed all the review comments. Please review the PR.

symbolist commented:
- [x] I tested this by turning on completion in LMS and verifying that the Aggregator objects are getting updated correctly when working through a course.
- [x] I read through the code
- [ ] ~~I checked for accessibility issues~~
- [ ] ~~Includes documentation~~

taranjeet commented:
@symbolist : I have addressed all the review comments and squashed all the commits into a single commit. Hence I am merging this PR.
Thanks for the review and the approval. :)
